### ISSUE TITLE
Weights of monkeypatched module get reset by a forward pass after being changed manually

### ISSUE THREAD

**Issue by Horse7354** [facebookresearch/higher#41]
Any computed update to the weights of a monkeypatched module (not with a differentiable optimizer) doesn't behave as expected:
```
data = torch.randn(32,10)  
module = torch.nn.Linear(10,10)  
fmodel = higher.monkeypatch(module,copy_initial_weights=True)  
fmodel.UpdateWeights() # some update to the weights, for a simple example: fmodel.weight = fmodel.weight * 2.
fmodel.forward(data) # weights are now reset as if fmodel.UpdateWeights() never happened
```
Edit: I've noticed that `fmodel.named_parameters()` and `fmodel._parameters` contain the correct updated parameters whereas `fmodel.parameters()` and `fmodel.fast_params` contain the old parameter values

**Comment by egrefen** [comment]
Thanks for posting this. I have a suspicion as to why this is happening, but need more info. Could you please post a full minimal working example, including what you expected to happen, what actually happened, and any relevant stack traces?

**Comment by Horse7354** [comment]
The above code works as an example with the substitution of UpdateWeights() to whatever operation on the weights:

```
data = torch.randn(32,2)
module = torch.nn.Linear(2,2)
fmodel = higher.monkeypatch(module)
w0=fmodel.weight.data
fmodel.weight = fmodel.weight * 2. # some update to the weights, for a simple example: fmodel.weight = fmodel.weight * 2.
w1=fmodel.weight.data
fmodel.forward(data) # weights are now reset as if fmodel.weight = fmodel.weight * 2. never happened
w2=fmodel.weight.data
print(w0,w1,w2) # see the proof
```

I'm not sure how to get the full stack trace but after patched_forward's setattr(https://github.com/facebookresearch/higher/blob/c6a79805b6d51c0571c04467a53cb1dc4e3754e4/higher/patch.py#L284), the parameter gets changed back to the original when this line is applied : https://github.com/facebookresearch/higher/blob/8c375871608d93e7424ff2d9e7009593a6def3e7/higher/patch.py#L211

**Comment by Horse7354** [comment]
Is there any potential workaround for the time being where I want "param" and potentially other meta parameters of module to get grad in the following example?:

```
module = torch.nn.Linear(2,2)
param = torch.tensor([2.],requires_grad=True)
fmodel = higher.monkeypatch(module) 

fmodel.weight = fmodel.weight * param

out = fmodel.forward(data)
loss = #some loss on out
loss.backward()
```


I tried having `module.load_state_dict(fmodel.state_dict())`  before the forward and then using the module, but the required parameters do not get grad. I also tried setting `._parameters['weight']` for module or fmodel in various ways but haven't got a working solution yet.

**Comment by egrefen** [comment]
Regarding a workaround, you can do `fmodel.parameters()`  to get the current state of the "fast" parameters. If you want to apply a (non in-place) function to a subset of them, do so on the returned iterable, forming a new iterable `new_params` with the same number of parameter tensors (i.e. including the unmodified ones), and then update the parameters by calling `fmodel.update_params(new_params)`. 

This isn't really a pattern I would encourage or will guarantee to support, but it should work for now.

**Comment by egrefen** [comment]
Okay, it looks like I have a solution. It wasn't simple but hopefully it's robust and will allow people to use patched modules even more naturally. Pull request coming up...

**Comment by egrefen** [comment]
Addressed in #43. Please re-open issue if it's not working for your use case.
